packages/editor/src/context/useResizableEditableTable.ts

```diff
@@ -0,0 +1,84 @@
+import type { ColumnDef, RowSelectionState, SortingState } from '@tanstack/react-table';
+import { getCoreRowModel, getSortedRowModel, useReactTable } from '@tanstack/react-table';
+
+import { useState, useEffect } from 'react';
+import { deepEqual } from '../utils/equals';
+interface UseResizableEditableTableProps<TData> {
+  data: TData[];
+  columns: ColumnDef<TData>[];
+  onChange: (change: TData[]) => void;
+  EMPTY_DATA_OBJECT: TData;
+}
+
+const useResizableEditableTable = <TData>({ data, columns, onChange, EMPTY_DATA_OBJECT }: UseResizableEditableTableProps<TData>) => {
+  const [sorting, setSorting] = useState<SortingState>([]);
+  const [rowSelection, setRowSelection] = useState<RowSelectionState>({});
+
+  const table = useReactTable({
+    data,
+    columns,
+    state: { sorting, rowSelection },
+    columnResizeMode: 'onChange',
+    columnResizeDirection: 'ltr',
+    enableRowSelection: true,
+    enableMultiRowSelection: false,
+    enableSubRowSelection: false,
+    onRowSelectionChange: setRowSelection,
+    onSortingChange: setSorting,
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    meta: {
+      updateData: (rowId, columnId, value) => {
+        const rowIndex = parseInt(rowId);
+        const updatedData = data.map((row, index) => {
+          if (index === rowIndex) {
+            return {
+              ...data[rowIndex]!,
+              [columnId]: value
+            };
+          }
+          return row;
+        });
+        if (!deepEqual(updatedData[updatedData.length - 1], EMPTY_DATA_OBJECT) && rowIndex === data.length - 1) {
+          onChange([...updatedData, EMPTY_DATA_OBJECT]);
+        } else {
+          onChange(updatedData);
+        }
+      }
+    }
+  });
+
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => !deepEqual(obj, EMPTY_DATA_OBJECT));
+
+      if (filteredData.length !== data.length) {
+        setRowSelection({});
+        onChange(filteredData);
+      }
+      return;
+    }
+  }, [data, EMPTY_DATA_OBJECT, onChange, rowSelection, table]);
+
+  const addRow = () => {
+    const newData = [...data];
+    newData.push(EMPTY_DATA_OBJECT);
+    onChange(newData);
+    setRowSelection({ [`${newData.length - 1}`]: true });
+  };
+
+  const removeRow = (index: number) => {
+    const newData = [...data];
+    newData.splice(index, 1);
+    if (newData.length === 0) {
+      setRowSelection({});
+    } else if (index === data.length - 1) {
+      setRowSelection({ [`${newData.length - 1}`]: true });
+    }
+    onChange(newData);
+  };
+
+  return { table, rowSelection, setRowSelection, addRow, removeRow };
```

Comment:
I would also return a function that has the state if the add row should be displayed: `showAddButton` or something the like.

Reply:
because then you don't have to do anything with this empty object outside of this object.